src/Span.ts

```diff
@@ -67,64 +67,8 @@ export function Span(config?: ISpanConfig) {
         const spanName = config.name || `${className}.${fnName}`;
         const spanType = config.type || className;
 
-        if (descriptor.value.constructor.name === 'AsyncFunction') {
-            descriptor.value = async function(...args: any[]) {
-                const span = activeApm.startSpan(spanName, spanType);
-
-                try {
-                    const ret = await original.apply(this, args);
-
-                    if (span) {
-                        span.end();
-                    }
-
-                    return ret;
-                } catch (err) {
-                    activeApm.captureError(err);
-                    const curTransaction = activeApm.currentTransaction;
-
-                    if (curTransaction) {
-                        curTransaction.result = 'error';
-                    }
-
-                    if (span) {
-                        span.end();
-                    }
-
-                    throw err;
-                }
-            };
-        } else {
-            descriptor.value = function(...args: any[]) {
-                const span = activeApm.startSpan(spanName, spanType);
-
-                try {
-                    const ret = original.apply(this, args);
-
-                    if (span) {
-                        if (ret && ret.then && ret.catch) {
-                            ret.then(() => span.end()).catch(() => span.end());
-                        } else {
-                            span.end();
-                        }
-                    }
-
-                    return ret;
-                } catch (err) {
-                    activeApm.captureError(err);
-                    const curTransaction = activeApm.currentTransaction;
-
-                    if (curTransaction) {
-                        curTransaction.result = 'error';
-                    }
-
-                    if (span) {
-                        span.end();
-                    }
-
-                    throw err;
-                }
-            };
-        }
+        descriptor.value = function(...args: any[]) {
+            return withSpan(original.bind(this), { name: spanName, type: spanType })(args);
```

Comment:
Calling `withSpan` with every function call will create a new function object every time the original function is called.

Reply:
I updated to reuse the same function in both `withSpan` and `Span`